src/adapt.c

```diff
@@ -16,6 +35,58 @@
 // #define VERBOSE 1
 
 
+/* Check if the given value is zero or not and return the
+ * correspondeding ADAPT_STATUS 
+ * */
+#define RETURN_ADAPT_STATUS(int) \
+    if (int) \
+        return ADAPT_OK; \
+    else \
+        return ADAPT_ERROR_WHILE_ADAPT;
+
+/* Free and set the given pointer to NULL 
+ * */
+#define FREE_AND_NULL(a) {\
+        free(a);\
+        a = NULL;\
+}
+
+/* Check pointer and then free it
+ * */
+#define CHECK_INIT_MALLOC_FREE(a) if(a)\
+    FREE_AND_NULL(a)
+
+/* unsusable stuff in the case that not enough memory is avaible 
+ * */
+#define CHECK_INIT_MALLOC(a) if( (a) == NULL) { \
+    config_destroy(&cfg); \
+    CHECK_INIT_MALLOC_FREE(c2conf_hashmap); \
+    CHECK_INIT_MALLOC_FREE(r2c_hashmap); \
+    CHECK_INIT_MALLOC_FREE(bids_hashmap); \
+    CHECK_INIT_MALLOC_FREE(function_stacks); \
+    CHECK_INIT_MALLOC_FREE(function_stack_sizes); \
+    CHECK_INIT_MALLOC_FREE(default_infos); \
+    CHECK_INIT_MALLOC_FREE(init_infos); \
+    CHECK_INIT_MALLOC_FREE(knob_offsets); \
+    return ENOMEM; \
+}
+
+
+/* Free a list
+ * static function would be nicer but is not possible for different
+ * structures 
+ * */
+#define FREE_LIST \
+    if (current->next){ \
+        current=current->next; \
+        while(current->next){ \
+            next=current->next; \
+            FREE_AND_NULL(current); \
+            current=next; \
+        } \
+        FREE_AND_NULL(current); \
+    }
+
```

Comment:
Still a macro instead of  a static function... And the initial current is not free'd